### DISCUSSION TITLE
Updating a document with a cluster-wide transaction and optimistic concurrency

### DISCUSSION THREAD

**Discussion by georgiosd** [ravendb/ravendb#15248] [General]
Hey everyone

I'm fascinated with how I'm still learning new things about Raven as new scenarios come up!

In this instance, I am using a cluster-wide transaction to make a sure an e-mail is unique but I also have to store a reference to that doc in another document. 

In the beginning I thought I can just patch the document and found that's not supported. Then I tried optimistic concurrency implicitly from session options or explicitly from `store(doc, changeVector, id)` and found that neither of those are supported either.

I ended up using a cluster-wide transaction for storing the e-mail doc and a single node session for storing the reference. Not my ideal scenario, because it's no longer an atomic operation.

I did bump into this old discussion but I didn't quite get the OP's suggestion for "incremental versions stored in cmp-xchg". 

Is there a better way to handle this scenario?

**Comment by grisha-kotler** [comment]
Hi,

Can't you store the referenced document using the cluster-wide transaction?

**Reply by georgiosd** [reply]
Yes, I already do that - I am asking about storing the document id in another document.

e.g. Let's say you have `UserEmailReference` which makes sure the email is unique and then `Entity` which has a collection of `UserEmailReference` ids. 

In my example, the `UserEmailReference` is created with the cluster-wide transaction, but `Entity` already exists and is modified.

**Comment by grisha-kotler** [comment]
Does the `Entity` store ALL the `UserEmailReferences`?
You can load the `Entity` and modify it. If someone modified it before you did, then you'll get a `ConcurrencyException` and you need to retry.

**Reply by georgiosd** [reply]
Doh! I was trying to be smart and load the object from the single-node session and then save it from the cluster session.

But I guess I was being silly in reality because the read is single node anyway, right?

**Reply by grisha-kotler** [reply]
You must use only one session type at a time.

**Reply by georgiosd** [reply]
Yes, I understand, though I expected "transferring" the object from one session to another using the change vector. 

The read part is a single node op in a cluster session too, right?

**Reply by ayende** [reply]
Yes, we always read from a single node.
Writing in a cluster wide transaction is always with optimistic concurrency (at the cluster level)
